hydra/hydra.go

```diff
@@ -192,14 +222,87 @@ func NewHydra(ctx context.Context, options Options) (*Hydra, error) {
 		go handleBootstrapStatus(hdCtx, bsCh)
 
 		hds = append(hds, hd)
+		hdPeerIds[hd.Host.ID()] = struct{}{}
 	}
+
 	fmt.Fprintf(os.Stderr, "\n")
 
 	for _, hd := range hds {
 		fmt.Fprintf(os.Stderr, "🆔 %v\n", hd.Host.ID())
 		for _, addr := range hd.Host.Addrs() {
 			fmt.Fprintf(os.Stderr, "🐝 Swarm listening on %v\n", addr)
 		}
+
+		// dial back hooks
+		evts := []interface{}{
+			new(event.EvtPeerIdentificationCompleted),
+		}
+
+		subs, err := hd.Host.EventBus().Subscribe(evts, eventbus.BufSize(256))
+		if err != nil {
+			return nil, fmt.Errorf("head could not subscribe to eventbus events; err: %w", err)
+		}
+		go func(hd *head.Head, subs event.Subscription) {
+			defer subs.Close()
+			for {
+				select {
+				case evt := <-subs.Out():
+					ev := evt.(event.EvtPeerIdentificationCompleted)
+					p := ev.Peer
+					// do not dial back our own heads
+					if _, ok := hdPeerIds[p]; ok {
+						continue
+					}
+					addrs := hd.Host.Peerstore().Addrs(p)
+
+					// dial back on quic if peer advertises a quic address
+					for _, a := range addrs {
+						if mafmt.QUIC.Matches(a) {
```

Comment:
I'm not sure off hand how this (or the TCP equivalent) interacts with circuit relay addresses, I don't know how many of those nodes are around but we may want to count them separately.

Reply:
I think this also catch p2p circuit addresses. Instead, I think we need to look at the _last_ protocol in the address.

Reply:
Really, we can probably just try and fail. The failure should be pretty fast.

Reply:
@Stebalien How are failures for QUIC fast ? We'll either have to wait for a timeout on our side or for the other end's Router to send us an ICMP error, right ?

Also, we should filter out relay addresses here before just dialing right ? Otherwise that would succeed and we might incorrectly judge that as dialable.

Reply:
@Stebalien 

Oh you mean the failure will be fast if there are all TCP addrs ? That makes sense.
However, we still need to filter out Relay addresses here, right ?

Reply:
Yes. But we'll ignore relay addresses as well if we simply don't enable the relay transport. Basically, if we construct two bare-bones hosts (maybe use the blank host?) and only enable one transport, all addresses that don't match that transport will fail.